src/control/input_3ds.cpp

```diff
@@ -0,0 +1,1162 @@
+    {
+        keys = this->m_keys;
+        keys2 = this->m_keys2;
+        key_max = PlayerControls::n_buttons;
+    }
+    else if(c == ControlsClass::Editor)
+    {
+        keys = this->m_editor_keys;
+        keys2 = this->m_editor_keys2;
+        key_max = EditorControls::n_buttons;
+    }
+    else if(c == ControlsClass::Hotkey)
+    {
+        keys = this->m_hotkeys;
+        keys2 = this->m_hotkeys2;
+        key_max = Hotkeys::n_buttons;
+    }
+    else
+    {
+        // BAD!
+        return true;
+    }
+
+    // minor switching algorithm to ensure that every button always has at least one key
+    // and no button ever has a non-unique key
+    // if a button's secondary key (including the current one) is the new key, delete it.
+    // if a button's primary key (excluding the current one) is the new key,
+    //     and it has a secondary key, overwrite it with the secondary key.
+    //     otherwise, replace it with the button the player is replacing.
+    for(size_t j = 0; j < key_max; j++)
+    {
+        if(keys2[j] == key)
+            keys2[j] = null_key;
+        else if(i != j && keys[j] == key)
+        {
+            if(keys2[j] != null_key)
+            {
+                keys[j] = keys2[j];
+                keys2[j] = null_key;
+            }
+            else
+                keys[j] = keys[i];
+        }
+    }
+
+    keys[i] = key;
+    return true;
+}
+
+bool InputMethodProfile_3DS::PollSecondaryButton(ControlsClass c, size_t i)
+{
+    // note: m_canPoll is initialized to false
+    InputMethodType_3DS* k = dynamic_cast<InputMethodType_3DS*>(this->Type);
+
+    if(!k)
+        return false;
+
+    uint32_t keys_held = hidKeysHeld();
+
+    int key;
+
+    for(key = 0; key < 32; key++)
+    {
+        if(BIT(key) & keys_held)
+            break;
+    }
+
+    // if didn't find any key, allow poll in future but return false
+    if(key == 32)
+    {
+        this->m_canPoll = true;
+        return false;
+    }
+
+    // if poll not allowed, return false
+    if(!this->m_canPoll)
+        return false;
+
+    // we will assign the key!
+    // reset canPoll for next time
+    m_canPoll = false;
+
+    // resolve the particular primary and secondary key arrays
+    int* keys;
+    int* keys2;
+    size_t key_max;
+
+    if(c == ControlsClass::Player)
+    {
+        keys = this->m_keys;
+        keys2 = this->m_keys2;
+        key_max = PlayerControls::n_buttons;
+    }
+    else if(c == ControlsClass::Cursor)
+    {
+        keys = nullptr;
+        keys2 = this->m_cursor_keys2;
+        key_max = CursorControls::n_buttons;
+    }
+    else if(c == ControlsClass::Editor)
+    {
+        keys = this->m_editor_keys;
+        keys2 = this->m_editor_keys2;
+        key_max = EditorControls::n_buttons;
+    }
+    else if(c == ControlsClass::Hotkey)
+    {
+        keys = this->m_hotkeys;
+        keys2 = this->m_hotkeys2;
+        key_max = Hotkeys::n_buttons;
+    }
+    else
+    {
+        // BAD!
```

Comment:
Possibly here you could add the log string (use log functions with `D_` prefix if you want to keep these entries printed during debug only, they never appear at release builds)

Reply:
I'll do this, thanks!